File: asm/src/main/java/dev/openrs2/asm/classpath/ClassPath.java
package dev.openrs2.asm.classpath;

import java.util.HashMap;
import java.util.Map;

import com.google.common.collect.ImmutableList;
import com.google.common.collect.ImmutableSet;
import dev.openrs2.asm.MemberDesc;
import dev.openrs2.asm.MemberRef;
import dev.openrs2.util.collect.DisjointSet;
import dev.openrs2.util.collect.ForestDisjointSet;
import org.objectweb.asm.tree.ClassNode;

public final class ClassPath {
	private final ClassLoader runtime;
	private final ImmutableList<Library> dependencies, libraries;
	private final Map<String, ClassMetadata> cache = new HashMap<>();

	public ClassPath(ClassLoader runtime, ImmutableList<Library> dependencies, ImmutableList<Library> libraries) {
		this.runtime = runtime;
		this.dependencies = dependencies;
		this.libraries = libraries;
	}

	public ImmutableList<Library> getLibraries() {
		return libraries;
	}

	public ImmutableList<ClassMetadata> getLibraryClasses() {
		var classes = ImmutableList.<ClassMetadata>builder();

		for (var library : libraries) {
			for (var clazz : library) {
				classes.add(get(clazz.name));
			}
		}

		return classes.build();
	}

	public ClassMetadata get(String name) {
		var metadata = cache.get(name);
		if (metadata != null) {
			return metadata;
		}

		for (var library : libraries) {
			var clazz = library.get(name);
			if (clazz != null) {
				metadata = new AsmClassMetadata(this, clazz, false);
				cache.put(name, metadata);
				return metadata;
			}
		}

		for (var library : dependencies) {
			var clazz = library.get(name);
			if (clazz != null) {
				metadata = new AsmClassMetadata(this, clazz, true);
				cache.put(name, metadata);
				return metadata;
			}
		}

		var reflectionName = name.replace('/', '.');

		Class<?> clazz;
		try {
			clazz = runtime.loadClass(reflectionName);
		} catch (ClassNotFoundException ex) {
			throw new IllegalArgumentException("Unknown class " + name);
		}

		metadata = new ReflectionClassMetadata(this, clazz);
		cache.put(name, metadata);
		return metadata;
	}

	public ClassNode getNode(String name) {
		for (var library : libraries) {
			var clazz = library.get(name);
			if (clazz != null) {
				return clazz;
			}
		}
		return null;
	}

	public DisjointSet<MemberRef> createInheritedFieldSets() {
		var disjointSet = new ForestDisjointSet<MemberRef>();
		var ancestorCache = new HashMap<ClassMetadata, ImmutableSet<MemberDesc>>();

		for (var library : libraries) {
			for (var clazz : library) {
				populateInheritedFieldSets(ancestorCache, disjointSet, get(clazz.name));
			}
		}

		return disjointSet;
	}

	private ImmutableSet<MemberDesc> populateInheritedFieldSets(Map<ClassMetadata, ImmutableSet<MemberDesc>> ancestorCache, DisjointSet<MemberRef> disjointSet, ClassMetadata clazz) {
		var ancestors = ancestorCache.get(clazz);
		if (ancestors != null) {
			return ancestors;
		}
		var ancestorsBuilder = ImmutableSet.<MemberDesc>builder();

		var superClass = clazz.getSuperClass();
		if (superClass != null) {
			var fields = populateInheritedFieldSets(ancestorCache, disjointSet, superClass);

			for (var field : fields) {
				var partition1 = disjointSet.add(new MemberRef(clazz.getName(), field));
				var partition2 = disjointSet.add(new MemberRef(superClass.getName(), field));
				disjointSet.union(partition1, partition2);
			}

			ancestorsBuilder.addAll(fields);
		}

		for (var superInterface : clazz.getSuperInterfaces()) {
			var fields = populateInheritedFieldSets(ancestorCache, disjointSet, superInterface);

			for (var field : fields) {
				var partition1 = disjointSet.add(new MemberRef(clazz.getName(), field));
				var partition2 = disjointSet.add(new MemberRef(superInterface.getName(), field));
				disjointSet.union(partition1, partition2);
			}

			ancestorsBuilder.addAll(fields);
		}

		for (var field : clazz.getFields()) {
			disjointSet.add(new MemberRef(clazz.getName(), field));
			ancestorsBuilder.add(field);
		}

		ancestors = ancestorsBuilder.build();
		ancestorCache.put(clazz, ancestors);
		return ancestors;
	}

	public DisjointSet<MemberRef> createInheritedMethodSets() {
		var disjointSet = new ForestDisjointSet<MemberRef>();
		var ancestorCache = new HashMap<ClassMetadata, ImmutableSet<MemberDesc>>();

		for (var library : libraries) {
			for (var clazz : library) {
				populateInheritedMethodSets(ancestorCache, disjointSet, get(clazz.name));
			}
		}

		return disjointSet;
	}

	private ImmutableSet<MemberDesc> populateInheritedMethodSets(Map<ClassMetadata, ImmutableSet<MemberDesc>> ancestorCache, DisjointSet<MemberRef> disjointSet, ClassMetadata clazz) {
		var ancestors = ancestorCache.get(clazz);
		if (ancestors != null) {
			return ancestors;
		}
		var ancestorsBuilder = ImmutableSet.<MemberDesc>builder();

		var superClass = clazz.getSuperClass();
		if (superClass != null) {
			var methods = populateInheritedMethodSets(ancestorCache, disjointSet, superClass);

			for (var method : methods) {
				var partition1 = disjointSet.add(new MemberRef(clazz.getName(), method));
				var partition2 = disjointSet.add(new MemberRef(superClass.getName(), method));
				disjointSet.union(partition1, partition2);
			}

			ancestorsBuilder.addAll(methods);
		}

		for (var superInterface : clazz.getSuperInterfaces()) {
			var methods = populateInheritedMethodSets(ancestorCache, disjointSet, superInterface);

			for (var method : methods) {
				var partition1 = disjointSet.add(new MemberRef(clazz.getName(), method));
				var partition2 = disjointSet.add(new MemberRef(superInterface.getName(), method));
				disjointSet.union(partition1, partition2);
			}

			ancestorsBuilder.addAll(methods);
		}

		for (var method : clazz.getMethods()) {
			disjointSet.add(new MemberRef(clazz.getName(), method));
			ancestorsBuilder.add(method);
		}

		ancestors = ancestorsBuilder.build();
		ancestorCache.put(clazz, ancestors);
		return ancestors;
	}
}

File: asm/src/main/java/dev/openrs2/asm/classpath/ClassPath.kt
package dev.openrs2.asm.classpath

import dev.openrs2.asm.MemberDesc
import dev.openrs2.asm.MemberRef
import dev.openrs2.asm.toBinaryClassName
import dev.openrs2.util.collect.DisjointSet
import dev.openrs2.util.collect.ForestDisjointSet
import org.objectweb.asm.Opcodes
import org.objectweb.asm.commons.Remapper
import org.objectweb.asm.tree.AbstractInsnNode
import org.objectweb.asm.tree.ClassNode

class ClassPath(
    private val runtime: ClassLoader,
    private val dependencies: List<Library>,
    val libraries: List<Library>
) {
    private val cache = mutableMapOf<String, ClassMetadata?>()

    /*
     * XXX(gpe): this is a bit of a hack, as it makes the asm module contain
     * some details that are only relevant in the deobfuscator. However, I
     * can't think of a better way of storing this state at the moment - ASM
     * doesn't have support for attaching arbitrary state to an
     * AbstractInsnNode. We need to persist the state across all of our
     * Transformers to avoid adding extraneous labels until the last possible
     * moment, which would confuse some of our analyses if added earlier.
     */
    val originalPcs = mutableMapOf<AbstractInsnNode, Int>()

    val libraryClasses: List<ClassMetadata>
        get() {
            val classes = mutableListOf<ClassMetadata>()
            for (library in libraries) {
                for (clazz in library) {
                    classes.add(get(clazz.name)!!)
                }
            }
            return classes
        }

    private inline fun computeIfAbsent(name: String, f: (String) -> ClassMetadata?): ClassMetadata? {
        if (cache.containsKey(name)) {
            return cache[name]
        }

        val clazz = f(name)
        cache[name] = clazz
        return clazz
    }

    operator fun get(name: String): ClassMetadata? = computeIfAbsent(name) {
        for (library in libraries) {
            val clazz = library[name]
            if (clazz != null) {
                return@computeIfAbsent AsmClassMetadata(this, clazz, false)
            }
        }

        for (library in dependencies) {
            val clazz = library[name]
            if (clazz != null) {
                return@computeIfAbsent AsmClassMetadata(this, clazz, true)
            }
        }

        val clazz = try {
            runtime.loadClass(name.toBinaryClassName())
        } catch (ex: ClassNotFoundException) {
            return@computeIfAbsent null
        }

        return@computeIfAbsent ReflectionClassMetadata(this, clazz)
    }

    fun getClassNode(name: String): ClassNode? {
        for (library in libraries) {
            val clazz = library[name]
            if (clazz != null) {
                return clazz
            }
        }

        return null
    }

    fun remap(remapper: Remapper) {
        for (library in libraries) {
            library.remap(remapper)
        }

        cache.clear()
    }

    fun createInheritedFieldSets(): DisjointSet<MemberRef> {
        return createInheritedMemberSets(ClassMetadata::fields, ClassMetadata::getFieldAccess, fields = true)
    }

    fun createInheritedMethodSets(): DisjointSet<MemberRef> {
        return createInheritedMemberSets(ClassMetadata::methods, ClassMetadata::getMethodAccess, fields = false)
    }

    private fun createInheritedMemberSets(
        getMembers: (ClassMetadata) -> List<MemberDesc>,
        getMemberAccess: (ClassMetadata, MemberDesc) -> Int?,
        fields: Boolean
    ): DisjointSet<MemberRef> {
        val disjointSet = ForestDisjointSet<MemberRef>()
        val ancestorCache = mutableMapOf<ClassMetadata, Set<MemberDesc>>()

        for (library in libraries) {
            for (clazz in library) {
                populateInheritedMemberSets(
                    getMembers,
                    getMemberAccess,
                    fields,
                    ancestorCache,
                    disjointSet,
                    get(clazz.name)!!
                )
            }
        }

        return disjointSet
    }

    private fun populateInheritedMemberSets(
        getMembers: (ClassMetadata) -> List<MemberDesc>,
        getMemberAccess: (ClassMetadata, MemberDesc) -> Int?,
        fields: Boolean,
        ancestorCache: MutableMap<ClassMetadata, Set<MemberDesc>>,
        disjointSet: DisjointSet<MemberRef>,
        clazz: ClassMetadata
    ): Set<MemberDesc> {
        val ancestors = ancestorCache[clazz]
        if (ancestors != null) {
            return ancestors
        }

        val ancestorsBuilder = mutableSetOf<MemberDesc>()

        for (superClass in clazz.superClassAndInterfaces) {
            val members =
                populateInheritedMemberSets(getMembers, getMemberAccess, fields, ancestorCache, disjointSet, superClass)

            for (member in members) {
                val access = getMemberAccess(clazz, member)
                if (access != null && (access and Opcodes.ACC_STATIC != 0 || fields)) {
                    continue
                }

                val partition1 = disjointSet.add(MemberRef(clazz.name, member))
                val partition2 = disjointSet.add(MemberRef(superClass.name, member))
                disjointSet.union(partition1, partition2)

                ancestorsBuilder.add(member)
            }
        }

        for (member in getMembers(clazz)) {
            disjointSet.add(MemberRef(clazz.name, member))
            ancestorsBuilder.add(member)
        }

        ancestorCache[clazz] = ancestorsBuilder
        return ancestorsBuilder
    }
}
